### PR TITLE
[AIR] Fix `ResourceChangingScheduler` not working with AIR

### LINKED ISSUE #26130: [AIR/Tune] `ResourceChangingScheduler` & AIR API
### What happened + What you expected to happen

`ResourceChangingScheduler` is not being tested with the new API and, after a cursory investigation, may not be usable in the first place. We should:

* Review the AIR APIs to ensure that it is possible to dynamically set resources/scaling config through the scheduler
* Add tests for ResourceChangingScheduler with AIR API and ensure that the scaling config gets updated accordingly

### Versions / Dependencies

master

### Reproduction script

-

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

This PR ensures that the new trial resources set by `ResourceChangingScheduler` are respected by the train loop logic by modifying the scaling config to match. Previously, even though trials had their resources updated, the scaling config was not modified which lead to eg. new workers not being spawned in the `DataParallelTrainer` even though resources were available.

In order to accomplish this, `ScalingConfigDataClass` is updated to allow equality comparisons with other `ScalingConfigDataClass`es (using the underlying PGF) and to create a `ScalingConfigDataClass` from a PGF.

Please note that this is an internal only change intended to actually make `ResourceChangingScheduler` work. In the future, `ResourceChangingScheduler` should be updated to operate on `ScalingConfigDataClass`es instead of PGFs as it is now. That will require a deprecation cycle.

## Related issue number

Closes https://github.com/ray-project/ray/issues/26130

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/trainable/session.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def trial_id(self) -> str:
 
     @property
     def trial_resources(self) -> Dict[str, float]:
```
Comment (Yard1) [inline python/ray/tune/trainable/session.py:58]: Fix type hint

**Comment by Yard1** [comment]
@sumanthratna FYI this PR will conflict with yours

File: python/ray/train/base_trainer.py [hunk 2]
```diff
@@ -178,7 +179,7 @@ def _validate_attributes(self):
         if not isinstance(self.scaling_config, dict):
             raise ValueError(
                 f"`scaling_config` should be an instance of `dict`, "
-                f"found {type(self.run_config)} with value `{self.run_config}`."
+                f"found {type(self.scaling_config)} with value `{self.scaling_config}`."
```
Comment (amogkam) [inline python/ray/train/base_trainer.py:182]: Nice good catch!

**Review by amogkam — COMMENTED** [review]
Thanks @Yard1! Changes to Trainer lgtm. 

Can we make sure to create an issue to track the remaining follow up items from the PR description and comments in the code?

File: python/ray/air/config.py [hunk 3]
```diff
@@ -283,9 +332,26 @@ class FailureConfig:
...
+
+    def __post_init__(self):
+        # Same check as in tune.run
+        if self.fail_fast and self.max_failures != 0:
+            raise ValueError("max_failures must be 0 if fail_fast=True.")
+
+        # Same check as in TrialRunner
+        if not (isinstance(self.fail_fast, bool) or self.fail_fast.upper() != "RAISE"):
+            raise ValueError(
+                "fail_fast must be one of {bool, 'raise'}. " f"Got {self.fail_fast}."
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/air/config.py:353]: We use lowercase in the docstring, so I figured we may as well be consistent here

**Comment by Yard1** [comment]
Will make issues once this is merged!

File: python/ray/air/config.py [hunk 4]
```diff
@@ -112,6 +128,39 @@ def as_placement_group_factory(self) -> "PlacementGroupFactory":
...
+
+        use_gpu = False
+        placement_strategy = pgf.strategy
+        resources_per_worker = None
+        num_workers = None
+
+        if worker_bundles:
+            first_bundle = worker_bundles[0]
+            if not all(bundle == first_bundle for bundle in worker_bundles[1:]):
+                raise ValueError()
```
Comment (krfricke) [inline python/ray/air/config.py:151]: Add better error message here

File: python/ray/air/tests/test_resource_changing.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        # and successfully save some model weights
+        checkpoint_dict = checkpoint.to_dict()
+        start_epoch = checkpoint_dict.get("epoch", -1) + 1
+
+    # wrap the model in DDP
+    for epoch in range(start_epoch, config["num_epochs"]):
+        checkpoint = Checkpoint.from_dict(dict(epoch=epoch))
+        session.report(
+            {"metric": config["metric"] * epoch, "epoch": epoch}, checkpoint=checkpoint
+        )
```
Comment (krfricke) [inline python/ray/air/tests/test_resource_changing.py:45]: I know we have tests for the resource changing scheduler in Tune. However, for the sanity of these tests, can we report `trial_resources.get("CPU")` here and later check that they are different for at least one trial?

**Review by krfricke — APPROVED** [review]
Looks good, left minor comments

File: python/ray/air/config.py (hunk 4 above)
Comment (Yard1) [reply python/ray/air/config.py]: Done
